allennlp/data/dataset_readers/encoder_decoder.py

```diff
@@ -0,0 +1,82 @@
+from typing import Dict
+import logging
+
+from overrides import overrides
+
+import tqdm
+
+from allennlp.common import Params
+from allennlp.common.checks import ConfigurationError
+from allennlp.data.dataset import Dataset
+from allennlp.data.dataset_readers import DatasetReader
+from allennlp.data.fields import TextField
+from allennlp.data.instance import Instance
+from allennlp.data.tokenizers import Tokenizer, WordTokenizer
+from allennlp.data.token_indexers import TokenIndexer, SingleIdTokenIndexer
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+START_SYMBOL = "@@START@@"
+END_SYMBOL = "@@END@@"
+
+@DatasetReader.register("encoder_decoder")
+class EncoderDecoderDatasetReader(DatasetReader):
+    """
+    Read a tsv file containing paired sequences, and create a dataset for an ``EncoderDecoder`` model.
+
+    Parameters
+    ----------
+    tokenizer : ``Tokenizer``, optional (default=``WordTokenizer()``)
+        Tokenizer to use to split the words in the input and output sequences.
+    input_token_indexers : ``Dict[str, TokenIndexer]``, optional
+                             (default=``{"tokens": SingleIdTokenizer("input_tokens")}``)
+        Indexers used to define input (source side) token representations.
+    output_token_indexers : ``Dict[str, TokenIndexer]``, optional
+                            (default=``{"tokens": SingleIdTokenizer("output_tokens")}``)
+        Indexers used to define output (target side) token representations.
+    """
+    def __init__(self, tokenizer: Tokenizer = None, input_token_indexers: Dict[str, TokenIndexer] = None,
+                 output_token_indexers: Dict[str, TokenIndexer] = None) -> None:
+        self._tokenizer = tokenizer or WordTokenizer()
+        self._input_token_indexers = input_token_indexers or {"tokens": SingleIdTokenIndexer("input_tokens")}
+        self._output_token_indexers = output_token_indexers or {"tokens": SingleIdTokenIndexer("output_tokens")}
+
+    @overrides
+    def read(self, file_path):
+        instances = []
+        with open(file_path, "r") as data_file:
+            logger.info("Reading instances from lines in file at: %s", file_path)
+            for line_num, line in enumerate(tqdm.tqdm(data_file)):
+                line = line.strip("\n")
+
+                if not line:
+                    continue
+
+                line_parts = line.split('\t')
+                if len(line_parts) != 2:
+                    raise ConfigurationError("Invalid line format: %s (line number %d)" % (line, line_num + 1))
+                source_sequence, target_sequence = line_parts
+                target_sequence = "%s %s %s" % (START_SYMBOL, target_sequence, END_SYMBOL)
```

Comment:
The target sequence probably doesn't need a start symbol, does it?  And you can get the behavior you want by adding an `end_token` in the tokenizer:

https://github.com/allenai/allennlp/blob/3b809fee6b37c6b99dd1fb5867b13e775e1f4b4d/allennlp/data/tokenizers/word_tokenizer.py#L36-L37

Probably just adding an assert in the constructor that tokenizer's `end_tokens` isn't empty is good enough.

Reply:
I think we need the start token because at test time that is the input at the first time step to the decoder. To keep things consistent, we need it while training as well.

Good to know about passing start and end tokens to the tokenizer. Will change that.

Reply:
Hmm...  Alternatively, you could just have a learned initial hidden state.  I guess having a start token with a learned embedding is _almost_ the same thing.  And this makes the API around decoding more consistent and cleaner.  Ok, yeah, you're right.  I'd still probably also learn the initial hidden state, but that's a separate issue.

Reply:
The initial hidden state of the decoder is supposed to be the final hidden state from the encoder, right?

Reply:
And regarding passing start and end symbols to the tokenizer, I'm using the same tokenizer for both source and target sequences, and we do not need start and end symbols in the source sequences. Would it make sense to change the tokenizer such that the `tokenize` method accepts start and end symbols, instead of the constructor?

Reply:
Yes, you're right on the hidden state.  Ignore my comment there.

Re: the tokenizer, for the semantic parser you almost certainly want different tokenizers for the input and the output, because the output is probably just space-separated, and trying to split it as if it were English would be wrong.  So, if your goal is to make a general piece of code that can be used for any encoder/decoder model (though I still think that may be trying to do too much), you'd want to have separate tokenizers, maybe defaulting the output tokenizer to be the same as the input tokenizer.

Reply:
I think this should be `input_sequence + DECODE_START_SYMBOL` and `target_sequence + END_SYMBOL` - I would say this is more standard. But Pradeep's method is equally valid, just maybe less standard. Also you are right that the hidden state is passed across from the encoder to the decoder.

Reply:
Sorry, ignore the above - there needs to be two target fields, one for the input to the decoder during training, which should be `DECODE_START + target_sequence` and one for the actual targets, which should be `target_sequence + END_SYMBOL`. Both of these should be optional model inputs.

Reply:
Matt, I agree that having separate tokenizers is a good idea. But I think the target tokenizer should default to a word tokenizer with start and end symbols, not the source tokenizer because even in the simplest case source and target should be tokenized differently.

Mark, how would adding a `DECODE_START_SYMBOL` at the end of the source sequence work? What would be the first input to the decoder then? I thought the whole point was to assign a symbol that can always be used as the first input to the decoder.

Reply:
Sorry, see my next comment.

Reply:
Hmm.. sounds like you're talking about targets and predictions. Is that the case? If so, what does it mean to have predictions as inputs to the model?

Reply:
No - at training time, the model has 2 inputs: the encoder_input and the decoder_input. Separately, there are targets, which are:

```
decoder_input = [START_TOKEN] + [words]
decoder_target = decoder_input[1:] + [END_TOKEN]
```
I meant to make both of these inputs to `forward` optional, because at test time, your input will be just the input_sequence and the decoder_input will be `[START_TOKEN]`. Does that make sense?

Reply:
Ah! I see what you mean. I understand the difference between decoder inputs and targets. But do we need them both as inputs to forward? Can we not just take the whole sequence [START_TOKEN] + [words] + [END_TOKEN] as the input and compute the two sequences? This should work as long as loss is computed against the right subset of the sequence, right?
(Thanks for bringing this up, though. I realized that my loss computation is not one-off. I should fix it. :))

Reply:
Yeah, that would work. My only thought regarding this would be that separating them might make doing things like incorporating teacher forcing easier, because you could just have an `if decoder_input is None: ` block which would sample from the model during training rather than using the `decoder_input`. If you don't separate them, then you'll have to pass a flag, but I guess that's a really trivial detail. Proceed as you wish 🚛

Reply:
Yes, I was going to do it with a flag. But I think the arguments to `forward` is an orthogonal issue because even while doing scheduled sampling, the method needs to get both the `decoder_inputs` and `predicted_outputs` and make a stochastic choice between the two.

In general, with these models, I cannot think of a case where `decoder_inputs = whole_sequence[:-1]` and `decoder_target = whole_sequence[1:]` does not hold true, and thus I do not think we should have them as separate inputs.

Reply:
Another argument for not splitting them: the direction we're going is to take a list of correct outputs, and doing constrained decoding (with different constraints at training and test time).  You don't have a per-token loss, just a loss at the end, and so you don't really want to think about things in just a "predict-the-next-token" sense.

I think the split decoder input/output makes it easy to implement the traditional seq2seq models efficiently, which is why Mark was suggesting that.  But our end goal is a bit different.